File: security-compliance-tw/testdata/sample-android/app/src/main/java/com/example/insecure/AuthManager.kt
package com.example.insecure

import android.content.Context
import android.util.Log
import java.security.MessageDigest
import java.util.Random
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * 刻意不安全的 fixture。不要抄。
 * 對應 MAST-STORAGE-001/002/004、MAST-CRYPTO-001/002/003、MAST-AUTH-001。
 */
class AuthManager(private val context: Context) {

    companion object {
        // MAST-STORAGE-004：硬編碼機密
        const val API_KEY = "FIXTURE-NOT-A-REAL-KEY-0000000000"
        const val JWT_SECRET = "FIXTURE-NOT-A-REAL-SECRET-000000"

        // MAST-CRYPTO-003：寫死的對稱金鑰與 IV
        private val AES_KEY = "0123456789abcdef".toByteArray()
        private val AES_IV = ByteArray(16)
    }

    // MAST-STORAGE-001：權杖明文寫進 SharedPreferences
    fun saveToken(token: String) {
        val prefs = context.getSharedPreferences("auth", Context.MODE_PRIVATE)
        prefs.edit().putString("access_token", token).apply()

        // MAST-STORAGE-002：敏感資料寫進日誌
        Log.d("AuthManager", "saved token=$token apiKey=$API_KEY")
    }

    // MAST-CRYPTO-001：已破解的雜湊演算法用於密碼
    fun hashPassword(password: String): String {
        val md = MessageDigest.getInstance("MD5")
        return md.digest(password.toByteArray()).joinToString("") { "%02x".format(it) }
    }

    // MAST-CRYPTO-002：非密碼學安全的亂數用於交談識別碼與 OTP
    fun newSessionId(): String = Random().nextLong().toString()

    fun newOtp(): Int = Random().nextInt(900000) + 100000

    // MAST-CRYPTO-003：固定金鑰 + 固定 IV
    fun encrypt(plain: ByteArray): ByteArray {
        val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(AES_KEY, "AES"), IvParameterSpec(AES_IV))
        return cipher.doFinal(plain)
    }

    // MAST-AUTH-001／003：用戶端自行裁決是否放行
    fun isAdmin(): Boolean =
        context.getSharedPreferences("auth", Context.MODE_PRIVATE)
            .getBoolean("is_admin", false)

    fun deleteUser(targetId: String) {
        if (isAdmin()) {
            Log.d("AuthManager", "deleting $targetId")
        }
    }
}

File: security-compliance-tw/testdata/sample-ios/App/AuthManager.swift
import CommonCrypto
import Foundation

/// 刻意不安全的 fixture。不要抄。
/// 對應 MAST-STORAGE-001/002/004、MAST-CRYPTO-001/002。
final class AuthManager {

    // MAST-STORAGE-004：硬編碼機密
    static let apiKey = "FIXTURE-NOT-A-REAL-KEY-0000000000"
    static let jwtSecret = "FIXTURE-NOT-A-REAL-SECRET-000000"

    // MAST-STORAGE-001：權杖明文寫進 UserDefaults
    func save(token: String) {
        UserDefaults.standard.set(token, forKey: "access_token")
        UserDefaults.standard.set(Self.apiKey, forKey: "api_key")

        // MAST-STORAGE-002：敏感資料寫進日誌
        print("saved token=\(token)")
        NSLog("apiKey=%@", Self.apiKey)
    }

    // MAST-CRYPTO-001：已破解的雜湊演算法
    func hashPassword(_ password: String) -> Data {
        var digest = [UInt8](repeating: 0, count: Int(CC_MD5_DIGEST_LENGTH))
        let data = Data(password.utf8)
        _ = data.withUnsafeBytes {
            CC_MD5($0.baseAddress, CC_LONG(data.count), &digest)
        }
        return Data(digest)
    }

    // MAST-CRYPTO-002：非密碼學安全的亂數
    func newSessionId() -> String {
        String(arc4random())
    }

    func newOtp() -> Int {
        Int(arc4random_uniform(900_000)) + 100_000
    }
}
